File: src/main/java/lk/tech/tgcontrollerclient/services/HttpRequests.java
package lk.tech.tgcontrollerclient.services;

import lk.tech.tgcontrollerclient.dto.ResultString;
import lk.tech.tgcontrollerclient.utils.BaseProvider;
import lombok.extern.slf4j.Slf4j;
import okhttp3.*;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Duration;

@Slf4j
public enum HttpRequests {

    INSTANCE;

    private final OkHttpClient client;

    HttpRequests() {
        this.client = new OkHttpClient.Builder()
                .connectTimeout(Duration.ofSeconds(10))
                .writeTimeout(Duration.ofSeconds(10))
                .readTimeout(Duration.ofSeconds(10))
                .retryOnConnectionFailure(true)
                .build();
    }

    // ==================================================
    //                   SEND TEXT
    // ==================================================
    public void sendText(String key, String command, String status) {
        try {
            String url = BaseProvider.httpUrl() +
                    "/api/v1/answer/text/" + key +
                    "?command=" + encode(command) +
                    "&status=" + encode(status);

            Request request = new Request.Builder()
                    .url(url)
                    .post(RequestBody.create(new byte[0]))
                    .build();

            client.newCall(request).enqueue(EMPTY_CALLBACK);

        } catch (Exception e) {
            log.error("[HTTP] sendText error", e);
        }
    }

    // ==================================================
    //                   SEND OBJECT
    // ==================================================
    public void sendObject(String key, String command, ResultString result) {
        try {
            String json = toJson(result);
            log.info("Sending object to server: {}", json);

            String url = BaseProvider.httpUrl() +
                    "/api/v1/answer/object/" + key +
                    "?command=" + encode(command) +
                    "&status=" + encode(result.getStatus());

            RequestBody body = RequestBody.create(
                    json.getBytes(StandardCharsets.UTF_8),
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url(url)
                    .post(body)
                    .build();

            client.newCall(request).enqueue(EMPTY_CALLBACK);

        } catch (Exception e) {
            log.error("[HTTP] sendObject error", e);
        }
    }

    // ==================================================
    //                   SEND IMAGE (PNG)
    // ==================================================
    public void sendImage(byte[] image, String key, String command, String status) {
        try {
            String url = BaseProvider.httpUrl() +
                    "/api/v1/answer/image/" + key +
                    "?command=" + encode(command) +
                    "&status=" + encode(status);

            RequestBody body = RequestBody.create(
                    image,
                    MediaType.parse("image/png")
            );

            Request request = new Request.Builder()
                    .url(url)
                    .post(body)
                    .build();

            client.newCall(request).enqueue(EMPTY_CALLBACK);

        } catch (Exception e) {
            log.error("[HTTP] sendImage error", e);
        }
    }

    // ==================================================
    //                     HELPERS
    // ==================================================

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    // Минимальная ручная JSON сериализация
    private static String toJson(ResultString r) {
        return """
               {"status":"%s","data":"%s"}
               """.formatted(
                escape(r.getStatus()),
                escape(r.getData())
        );
    }

    private static String escape(String s) {
        if (s == null) return "";
        StringBuilder sb = new StringBuilder(s.length());

        for (char c : s.toCharArray()) {
            switch (c) {
                case '\\' -> sb.append("\\\\");
                case '"'  -> sb.append("\\\"");
                case '\b' -> sb.append("\\b");
                case '\f' -> sb.append("\\f");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                default -> {
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else sb.append(c);
                }
            }
        }

        return sb.toString();
    }

    // Пустой callback чтобы не спамить
    private static final Callback EMPTY_CALLBACK = new Callback() {
        @Override public void onFailure(Call call, IOException e) {
            log.error("[HTTP] Request failed: {}", e.getMessage());
        }

        @Override public void onResponse(Call call, Response response) {
            response.close();
        }
    };
}

File: services/http/http_requests.go
package http

import (
	"bytes"
	"encoding/json"
	"fmt"
	"mime/multipart"
	"net/http"
	"net/url"
	"tg-controller-client/dto"
	"tg-controller-client/utils"
	"time"
)

type HttpRequests struct {
	client *http.Client
}

var HR = NewHttpRequests()

func NewHttpRequests() *HttpRequests {
	return &HttpRequests{
		client: &http.Client{
			Timeout: 10 * time.Second,
		},
	}
}

// ==================================================
//                 SEND TEXT
// ==================================================

func (h *HttpRequests) SendText(key, command, status string) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("[HTTP] sendText panic:", r)
		}
	}()

	base := utils.HttpUrl()

	u := fmt.Sprintf(
		"%s/api/v1/answer/%s?command=%s&status=%s",
		base,
		key,
		url.QueryEscape(command),
		url.QueryEscape(status),
	)

	req, err := http.NewRequest("POST", u, bytes.NewBuffer([]byte{}))
	if err != nil {
		fmt.Println("[HTTP] SendText: NewRequest error:", err)
		return
	}

	_, err = h.client.Do(req)
	if err != nil {
		fmt.Println("[HTTP] SendText error:", err)
	}
}

// ==================================================
//                 SEND OBJECT (JSON)
// ==================================================

func (h *HttpRequests) SendObject(key, command string, result dto.ResultString) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("[HTTP] sendObject panic:", r)
		}
	}()

	jsonData, err := json.Marshal(result)
	if err != nil {
		fmt.Println("[HTTP] JSON marshal error:", err)
		return
	}

	fmt.Println("Sending object to server:", string(jsonData))

	base := utils.HttpUrl()
	u := fmt.Sprintf(
		"%s/api/v1/answer/%s?command=%s&status=%s",
		base,
		key,
		url.QueryEscape(command),
		url.QueryEscape(result.Status),
	)

	req, err := http.NewRequest(
		"POST",
		u,
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		fmt.Println("[HTTP] SendObject: NewRequest error:", err)
		return
	}

	req.Header.Set("Content-Type", "application/json")

	_, err = h.client.Do(req)
	if err != nil {
		fmt.Println("[HTTP] SendObject error:", err)
	}
}

// ==================================================
//                 SEND IMAGE (PNG)
// ==================================================

func (h *HttpRequests) SendImagesMultipart(images [][]byte, key, command, status string) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("[HTTP] SendImagesMultipart panic:", r)
		}
	}()

	base := utils.HttpUrl()

	u := fmt.Sprintf(
		"%s/api/v1/answer/%s?command=%s&status=%s",
		base,
		key,
		url.QueryEscape(command),
		url.QueryEscape(status),
	)

	var body bytes.Buffer
	writer := multipart.NewWriter(&body)

	// добавляем файлы
	for i, img := range images {
		fileName := fmt.Sprintf("image_%d.png", i)

		part, err := writer.CreateFormFile("files", fileName)
		if err != nil {
			fmt.Println("[HTTP] CreateFormFile error:", err)
			return
		}

		_, err = part.Write(img)
		if err != nil {
			fmt.Println("[HTTP] writing image error:", err)
			return
		}
	}

	// закрываем multipart writer (обязательно)
	writer.Close()

	req, err := http.NewRequest("POST", u, &body)
	if err != nil {
		fmt.Println("[HTTP] NewRequest error:", err)
		return
	}

	// самое главное — тип multipart/form-data
	req.Header.Set("Content-Type", writer.FormDataContentType())

	_, err = h.client.Do(req)
	if err != nil {
		fmt.Println("[HTTP] SendImagesMultipart error:", err)
	}
}
